### ISSUE TITLE
TabNavigator didFocus event triggers before animation

### ISSUE THREAD

**Issue by kmwillia** [react-navigation/react-navigation#3537]
### Current Behavior
I have a screen in a TabNavigator which subscribes to the didFocus event.  Until the event fires, the component displays a loading icon.  Once the didFocus event fires, the component updates its state, and renders something slow to render.  Per the [API Reference](https://reactnavigation.org/docs/navigation-prop.html#addlistener-subscribe-to-updates-to-navigation-lifecycle), the didFocus event should trigger once the transition is complete.  When swiping, this appears to be the case, and the event fires after the animation completes.  However, when tapping the Tab to navigate, the event appears to trigger before the animation, causing a noticeable delay before the transition begins.

The component I'm using is as follows, and the project can be found in [here](https://github.com/kmwillia/rn-nav-did-focus).
``` javascript
const exampleList = new Array(1000).fill(undefined).map((value, index) => ({ key: index, label: `List Item ${index}` }));

class LazyScreen extends Component {
  // Should render a placeholder component until it receives didFocus event,
  // then it should display a slow to render list
  state = { loaded: false };

  constructor(...args) {
    super(...args);
    this.props.navigation.addListener('didFocus', () => {
      if(!this.state.loaded) {
        this.setState({ loaded: true });
      }
    });
  }

  render() {
    return this.state.loaded
      ? <ScrollView style={styles.slowScreen}>
          {exampleList.map(item => <Text key={item.key}>{item.label}</Text>)}
        </ScrollView>
      : <ActivityIndicator size={'large'} style={{ padding: 16 }} />;
  }
}
```
Swipe:
![swipe](https://user-images.githubusercontent.com/3580007/36328292-97ea6ffa-132f-11e8-9223-27921f62f4c0.gif)
Tap:
![tap](https://user-images.githubusercontent.com/3580007/36328303-9ea02402-132f-11e8-945f-30eb5dab3876.gif)

### Expected Behavior
Swiping and tapping should trigger the didFocus event after the transition is completed.

### How to reproduce
Example code can be seen [here](https://github.com/kmwillia/rn-nav-did-focus).  And the Snack is [here](https://snack.expo.io/B1hJOTEvz). 
 Tabs Zero, One, and Two are simple static screens.  The last Tab, Lazy, renders a placeholder, or a long list of items dependent on whether the didFocus event has fired.
- Have a TabNavigator with at least two screens.
- The second screen should subscribe to the didFocus event, and rely on it to switch from rendering a quick placeholder, to a slower component.
- Compare swiping and tapping to navigate to the second screen.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 1.0.3
| react-native     | 0.53.0
| node             | 8.9.4
| npm or yarn      | npm

**Comment by 7772** [comment]
+1

**Comment by yehudahkay** [comment]
+1

**Comment by brentvatne** [comment]
appears to be resolved on 2.0 https://snack.expo.io/SJjqPL06z

**Comment by Skylnt** [comment]
thanks, @brentvatne
